tests/test_download_images.py

```diff
@@ -0,0 +1,284 @@
+import unittest
+from unittest.mock import patch, MagicMock, mock_open, call
+import pandas as pd
+import os
+import io
+import shutil
+from io import BytesIO
+from PIL import Image
+import requests
+import base64
+from cautiousrobot.__main__ import download_images, main
+
+class TestDownload(unittest.TestCase):
+    def setUp(self):
+        self.DUMMY_DATA = pd.DataFrame(data={
+            "filename": ["test_file1", "test_file2"],
+            "file_url": ["http://test_url1.com/image.jpg", "http://test_url2.com/image.jpg"],
+            "subfolder": ["test_subfolder1", "test_subfolder2"]
+        })
+        self.IMG_DIR = "test_dir"
+        self.LOG_FILEPATH = "test_log_path.jsonl"
+        self.ERROR_LOG_FILEPATH = "test_error_log_path.jsonl"
+        self.DOWNSAMPLE_PATH = "test_downsample_dir"
+        self.DOWNSAMPLE_SIZE = 100
+
+        os.makedirs(self.IMG_DIR, exist_ok=True)
+        os.makedirs(self.DOWNSAMPLE_PATH, exist_ok=True)
+
+    def tearDown(self):
+        shutil.rmtree(self.IMG_DIR, ignore_errors=True)
+        shutil.rmtree(self.DOWNSAMPLE_PATH, ignore_errors=True)
+        if os.path.exists(self.LOG_FILEPATH):
+            os.remove(self.LOG_FILEPATH)
+        if os.path.exists(self.ERROR_LOG_FILEPATH):
+            os.remove(self.ERROR_LOG_FILEPATH)
+
+    @patch('requests.get')
+    def test_response_exception(self, get_mock):
+        get_mock.side_effect = requests.exceptions.RequestException
+        download_images(self.DUMMY_DATA, self.IMG_DIR, self.LOG_FILEPATH, self.ERROR_LOG_FILEPATH)
+        for filename in self.DUMMY_DATA['filename']:
+            self.assertFalse(os.path.isfile(f"{self.IMG_DIR}/{filename}"))
+
+    @patch('requests.get')
+    def test_successful_download(self, get_mock):
+        mock_response = MagicMock()
+        mock_response.status_code = 200
+        mock_response.raw = BytesIO(b"fake_image_data")
+        get_mock.return_value = mock_response
+
+        download_images(self.DUMMY_DATA, self.IMG_DIR, self.LOG_FILEPATH, self.ERROR_LOG_FILEPATH)
+
+        for filename in self.DUMMY_DATA['filename']:
+            self.assertTrue(os.path.isfile(f"{self.IMG_DIR}/{filename}"))
+
+    def retry_test_template(self, get_mock, status_code):
+        mock_response_retry = MagicMock()
+        mock_response_retry.status_code = status_code
+        mock_response_success = MagicMock()
+        mock_response_success.status_code = 200
+        mock_response_success.raw = BytesIO(b"fake_image_data")
+        get_mock.side_effect = [mock_response_retry] * 2 + [mock_response_success]
+
+        download_images(self.DUMMY_DATA, self.IMG_DIR, self.LOG_FILEPATH, self.ERROR_LOG_FILEPATH, retry=3)
+
+        for filename in self.DUMMY_DATA['filename']:
+            self.assertTrue(os.path.isfile(f"{self.IMG_DIR}/{filename}"))
+
+    @patch('requests.get')
```

Comment:
Can you replace these with a loop over retry codes (I think you could do something similar to @thompsonmj's update on the [sum-buddy `hasher` test](https://github.com/Imageomics/sum-buddy/pull/16/files#diff-d135cf68fe42d10ee2c17c30b1f38af17466b46368aa593f77a8fa45c0e61d73)).